base/parse.jl

```diff
@@ -216,22 +215,39 @@ end
 
 ## string to float functions ##
 
-tryparse(::Type{Float64}, s::String) = ccall(:jl_try_substrtod, Nullable{Float64}, (Ptr{UInt8},Csize_t,Csize_t), s, 0, sizeof(s))
-tryparse(::Type{Float64}, s::SubString{String}) = ccall(:jl_try_substrtod, Nullable{Float64}, (Ptr{UInt8},Csize_t,Csize_t), s.string, s.offset, s.endof)
-
-tryparse(::Type{Float32}, s::String) = ccall(:jl_try_substrtof, Nullable{Float32}, (Ptr{UInt8},Csize_t,Csize_t), s, 0, sizeof(s))
-tryparse(::Type{Float32}, s::SubString{String}) = ccall(:jl_try_substrtof, Nullable{Float32}, (Ptr{UInt8},Csize_t,Csize_t), s.string, s.offset, s.endof)
-
+function tryparse(::Type{Float64}, s::String)
+    hasvalue, val = ccall(:jl_try_substrtod, Tuple{Bool, Float64},
+                          (Ptr{UInt8},Csize_t,Csize_t), s, 0, sizeof(s))
+    hasvalue ? Some(val) : null
+end
+function tryparse(::Type{Float64}, s::SubString{String})
+    hasvalue, val = ccall(:jl_try_substrtod, Tuple{Bool, Float64},
+                          (Ptr{UInt8},Csize_t,Csize_t), s.string, s.offset, s.endof)
+    hasvalue ? Some(val) : null
+end
+function tryparse(::Type{Float32}, s::String)
+    hasvalue, val = ccall(:jl_try_substrtof, Tuple{Bool, Float32},
+                          (Ptr{UInt8},Csize_t,Csize_t), s, 0, sizeof(s))
+    hasvalue ? Some(val) : null
+end
+function tryparse(::Type{Float32}, s::SubString{String})
+    hasvalue, val = ccall(:jl_try_substrtof, Tuple{Bool, Float32},
+                          (Ptr{UInt8},Csize_t,Csize_t), s.string, s.offset, s.endof)
+    hasvalue ? Some(val) : null
+end
 tryparse(::Type{T}, s::AbstractString) where {T<:Union{Float32,Float64}} = tryparse(T, String(s))
 
-tryparse(::Type{Float16}, s::AbstractString) = convert(Nullable{Float16}, tryparse(Float32, s))
+function tryparse(::Type{Float16}, s::AbstractString)
```

Comment:
The replacement is much more verbose than the original because the shorter form:
```julia
convert(Union{Some{Float16}, Null}, tryparse(Float32, s))
```
does not work without a `convert` method like this:
```julia
convert(::Type{Union{Some{T}, Null}}, x::Some) where {T} = convert(Some{T}, x)
```

However, adding that `convert` method triggers a failure in the `ambiguous` test:
```julia
detect_unbound_args in Base and Core: Test Failed
  Expression: need_to_handle_undef_sparam == Set()
   Evaluated: Set(Method[convert(::Type{Union{Null, Some{T}}}, x::Some) where T in Main at REPL[1]:1]) == Set{Any}()
```

Does this indicate a real problem with that `convert` method? Or is it OK to add it to the already present exceptions in the test?

Reply:
`Nulls.jl` defines

```julia
Base.convert(::Type{Union{T, Null}}, x) where {T} = convert(T, x)
```

Would it make sense to just use this method?

Reply:
Unfortunately, it triggers the same test failure. But it would indeed make sense to define that method, which is more general.

Reply:
You can add that method to the list of allowed exceptions, since the necessary ambiguity resolver is present (we just have limited logic implemented there to compute and detect those intersections).